plugin/pkg/admission/podpreset/exists/admission.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exists
+
+import (
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/discovery"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	scadmission "github.com/kubernetes-incubator/service-catalog/pkg/apiserver/admission"
+)
+
+const (
+	// PluginName is name of admission plug-in
+	PluginName = "ServiceCatalogPodPresetExists"
+	// SupportedPodPresetVersion is the GroupVersion for what we require for
+	// creating PodPresetTemplates in bindings.
+	SupportedPodPresetVersion = "settings.k8s.io/v1alpha1"
+)
+
+func init() {
+	admission.RegisterPlugin(PluginName, func(io.Reader) (admission.Interface, error) {
+		return NewExists()
+	})
+}
+
+// exists is an implementation of admission.Interface.
+// It enforces that if a binding is being defined with a PodPresetTemplate
+// that the cluster actually has support for it.
+type exists struct {
+	*admission.Handler
+	discovery         discovery.DiscoveryInterface
+	sync.Mutex        // protects podPresetsChecked and podPresetsExists
+	podPresetsChecked bool
+	podPresetsExists  bool
+}
+
+var _ = scadmission.WantsKubeClientSet(&exists{})
+
+func (e *exists) supportsPodPresets() bool {
+	e.Lock()
+	defer e.Unlock()
+	return e.podPresetsExists
+}
+
+// checkClusterForPodPresets uses discovery to find if the cluster has the
+// right version of settings that supports PodPresets. It will return true
+// if we have successfully queried the Discovery API and hence we know
+// for sure if the cluster supports PodPresets or not.
+// In that case, it also sets podPresetsChecked to true and podPresetsExists
+// to true if the cluster supports PodPresets.
+func (e *exists) checkClusterForPodPresets() bool {
```

Comment:
as an accessory function that only runs once, I'd prefer if this was moved down under `SetKubeClientSet`.